Refactor get_builder_dt legacy to be explicit

nicain opened the pull request:
## Motivation

Resolves #258 

Debugging and developing legacy files (for example #275) is difficult because the code paths that legacy/map.py::get_builder_dt.py can take are very complicated.  This commit refactors this function, to make its behavior explicit and (relatively) easy.

## How to test the behavior?
```
make test
```

## Checklist

- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR description clearly describes problem and the solution?
- [x] Is your contribution compliant with our coding style ? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [x] Have you included the relevant issue number using `#XXX` notation where `XXX` is the issue number ?

nicain committed:
Refactor get_builder_dt legacy to be explicit

PEP8 compliance

PEP8 compliance

Add docstring. flake8 complexity skip. meaningful RuntimeError

pep8speaks commented:
Hello @nicain! Thanks for updating the PR.



Cheers ! There are no PEP8 issues in this Pull Request. :beers: 

##### Comment last updated on December 04, 2017 at 20:08 Hours UTC

jcfr commented:
@nicain you could also run `flake8` locally to find out about the error.

When addressed, please consider squashing all the commits together.

nicain commented:
@jcfr I think I have finally figured out how to do that :)  Ill add you as a reviewer too.

nicain commented:
@jcfr OK, flake is complaining about "./src/pynwb/legacy/map.py:26:5: C901 'TypeMapLegacy.get_builder_dt' is too complex (23)"
I want to tell it, well, you should have seen it before!  Any thoughts on the refactor?  at least its PEP8 compliant...

BTW running flake8 on my dev machine gives > 6000 problems.  Is there any way to only run it on the diff?

jcfr commented:
> @jcfr OK, flake is complaining about "./src/pynwb/legacy/map.py:26:5: C901 'TypeMapLegacy.get_builder_dt' is too complex (23)"

If the complexity is legitimate, you can add a comment like this one: 

```
def get_builder_dt(self, builder):  # noqa: C901
```


> running flake8 on my dev machine gives > 6000 problems

This is strange.

Could you confirm you do something like this:

```
cd /home/jcfr/Projects/pynwb
workon pyenv-py35
flake8
# => No error report
```

Could you also check you have the latest version of flake8 ?

jcfr commented on src/pynwb/legacy/map.py:
```diff
@@ -25,54 +25,93 @@ class TypeMapLegacy(TypeMap):
 
     def get_builder_dt(self, builder):
```
Add a docstring explaining what the function do. One sentence is good enought

jcfr replied:
Add a docstring explaining what the function do. One sentence is good enough

jcfr commented on src/pynwb/legacy/map.py:
```diff
@@ -25,54 +25,93 @@ class TypeMapLegacy(TypeMap):
+            attrs = builder.attributes
+            ndt = attrs.get('neurodata_type')
+
+            if ndt == 'Module':
```
To reduce the complexity, consider adding a function named `_get_builder_dt_from_neurodata_type(neurodata_type, attrs) ?`

nicain replied:
I want to keep the functionality as parallel to src/form/build/map.py::get_builder_dt, so I would prefer to use noqa: C901 unless you really think it is unreable w/o a refactor.

jcfr commented on src/pynwb/legacy/map.py:
```diff
@@ -25,54 +25,93 @@ class TypeMapLegacy(TypeMap):
+                    return 'ImageSeries'
+                else:
+                    return ancestry
+            elif ndt == 'Custom':
```
And may be a function named `def get_builder_type_from_custom_neurodata_type` or similar name.

That said, it is also nice to have all type handling in one function, and add the `noqa` exception can be legitimate.

nicain replied:
same as above

jcfr commented on src/pynwb/legacy/map.py:
```diff
@@ -25,54 +25,93 @@ class TypeMapLegacy(TypeMap):
                 else:
-                    return 'PlaneSegmentation'
-
+                    raise Exception
```
Avoid generic Exception. Choose from an existing specialized exception `RuntimeError` or similar . Make also sure to associate a descriptive message

nicain commented:
Addressing @jcfr:
- [x]: Using flake8 locally before push (failure in my dev env due to my .history dir not being --excluded)
- [x]: Using # noqa: C901 to skip complexity check (complexity is legitimate)
- [x]: Added docstring
- [x]: Exception --> RuntimeError
- [x]: Squashed commit history

jcfr commented on src/pynwb/legacy/map.py:
```diff
@@ -23,56 +23,101 @@ def source_gettr(self, builder, manager):
+        For a given builder, return the neurodata_type.  In this legacy
+        TypeMap, the builder may have out-of-spec neurodata_type; this
+        function coerces this to a 2.0 compatible version.
+        '''
```
One last nitpick: Prefer `"""` instead of `'''`

jcfr replied:
Or for consistency, we can update all docstring in a later commit

jcfr reviewed (dismissed):
Assuming CI returns green, LGTM

nicain commented:
@dorukozturk @jcfr Need a little help here.  Travis is complaining about:

> The command "pip install -U scikit-ci-addons" failed and exited with 127 during .

, and appveyor hasen't even started after 16 hours.  Is there something I did, or a misconfiguration somewhere?

dorukozturk commented:
@nicain I can take a look at travis. 

I don't know about appveyor though. It looks like lot of things got queued. I canceled some of them, I hope that helps. 

BTW definitely not something you did. :)

jcfr commented:
>  travis

Probably related to https://blog.travis-ci.com/2017-11-21-xcode8-3-default-image-announce

I submitted a PR on a different project to try something. Let's monitor how it goes. See https://github.com/scikit-build/scikit-ci-addons/pull/62/commits/ee5143dc040af6dcc675afc21c5149e8f452e304

dorukozturk commented:
@nicain Let's merge this. There is a major issue with pyenv on travis right now (see [this](https://github.com/travis-ci/travis-ci/issues/8363)). I will be watching the travis issue and fix it later.

nicain commented:
@jcfr Great. I just rebased, incorporating https://github.com/NeurodataWithoutBorders/pynwb/pull/292.  Assuming all but travis passes, please merge.

nicain commented:
@dorukozturk I clicked in on the appveyor check, and it reports jobs that have been queued for 5 days and targeting commit f9604032, not bf386b3 (the current head to be merged in). I wonder if that frozen state that appveyor helped you fix last week is somehow not getting applied to this PR.  Either way, it looks like this check needs to be restarted.  Can you take a look?

dorukozturk commented:
@nicain Appveyor let us do 1 build at a time. Since I merged another branch, it was running the master first. Your PR is currently running see [this](https://ci.appveyor.com/project/NeurodataWithoutBorders/pynwb/build/1.0.587).
